File: my_cpp_pkg/src/led_panel.cpp
#include "rclcpp/rclcpp.hpp"
#include "my_robot_interfaces/msg/led_state_array.hpp"
#include "my_robot_interfaces/srv/set_led.hpp"

using namespace std::chrono_literals;
using namespace std::placeholders;


class LedPanelNode : public rclcpp::Node
{
public:
    LedPanelNode() : rclcpp::Node("led_panel_state")
    {
        this->declare_parameter("led_state", std::vector<int64_t>{0,0,0});
        led_state_ = this->get_parameter("led_state").as_integer_array();
        publisher_ = this->create_publisher<my_robot_interfaces::msg::LedStateArray>("led_panel_status", 10);
        timer_ = this->create_wall_timer(
            1s,
            std::bind(&LedPanelNode::publishLedPanelStatus, this)
        );
        service_ = this->create_service<my_robot_interfaces::srv::SetLed>(
            "set_led",
            std::bind(&LedPanelNode::callbackSetLed, this, _1, _2));
        RCLCPP_INFO(this->get_logger(), "Led Pane Node has been started.");

    }

private:
    rclcpp::Publisher<my_robot_interfaces::msg::LedStateArray>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    std::vector<int64_t> led_state_;
    rclcpp::Service<my_robot_interfaces::srv::SetLed>::SharedPtr service_;

    void publishLedPanelStatus()
    {
        auto msg = my_robot_interfaces::msg::LedStateArray();
        msg.led_states = led_state_;
        publisher_->publish(msg);
    }
    void callbackSetLed(
        const std::shared_ptr<my_robot_interfaces::srv::SetLed::Request> request,
        std::shared_ptr<my_robot_interfaces::srv::SetLed::Response> response)
    {
      int64_t led_number = request->led_number;
      int64_t state = request->state;
      if(led_number >= (int64_t)led_state_.size() || led_number < 0)
      {
        response->success = false;
        return;
      }
      if (state!=0 && state!=1)
      {
        response->success = false;
        return;
      }
      led_state_.at(led_number) = state;
      response->success = true;
      publishLedPanelStatus();
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<LedPanelNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: my_py_pkg/my_py_pkg/led_panel.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from my_robot_interfaces.msg import LedStateArray
from my_robot_interfaces.srv import SetLed 

class LedPanelNode(Node):
    def __init__(self):
        super().__init__("led_panel")
        self.declare_parameter("led_states", [0, 0, 0])
        self.led_states_ = self.get_parameter("led_states").value
        self.led_panel_server_ = self.create_service(
            SetLed, "set_led", self.callback_set_led)
        self.get_logger().info("LED Panel Service has been started.") 
        self.led_states_publisher_ = self.create_publisher(
            LedStateArray, "led_panel_state", 10)
        self.led_states_timer_ = self.create_timer(
            5.0, self.publish_led_states)
        self.get_logger().info("LED PanelNode has been started.")

    def publish_led_states(self):
        msg = LedStateArray()
        msg.led_states = self.led_states_
        self.led_states_publisher_.publish(msg)
        self.get_logger().info(f"Published LED states: {self.led_states_}")

    def callback_set_led(self,request: SetLed.Request, response: SetLed.Response):
        led_number = request.led_number
        state = request.state

        if led_number>=len(self.led_states_) or led_number<0:
            response.success = False
            return response
        if state not in [0,1]:
            response.success = False
            return response
        self.led_states_[led_number] = state
        self.publish_led_states()        
        response.success = True
        return response



def main(args=None):
    rclpy.init(args=args)
    node = LedPanelNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()
